### DISCUSSION TITLE
Using a secret manager

### DISCUSSION THREAD

**Discussion by LilTrublMakr** [cross-seed/cross-seed#738] [Q&A]
Hello all,

I am trying to integrate a secret manager ([Infisical](https://infisical.com)) to grab the API keys of the *arrs. I am trying to do this in case I need to change the API keys, I can change them in one place and the rest of my services should update automatically.

I am thinking that the request will need to be made in javascript since that is what the config file is using. It is not going well but I don't really know what I am doing.

Has anyone successfully been able to integrate a call to pull secrets from a secret manager for the *arr API values?

**Comment by zakkarry** [comment]
There is an [infisical SDK (`npm install infisical`)](https://www.npmjs.com/package/infisical) although it seems to be deprecated - I'm not sure if it's still functional or not with the latest version of infisical. I'd assume you _SHOULD_ be running the latest version of any secret/password manager.

The code looks fairly straightforward, though, if the package still works.

**Reply by LilTrublMakr** [reply]
Would there be a possibility that the npm package `dotenv` could be included in the docker image? I am able to have the secrets generated in the host machine and the .env file would generate in to the config folder. I am unable to sh in to the container and install it manually because of permission issues that I am unsure how to get around.

**Reply by zakkarry** [reply]
I'm not sure what you mean `because of permissions issue` - you most certainly should be able to use `docker exec -it cross-seed /bin/sh` - i do this regularly.

**Reply by LilTrublMakr** [reply]
```
root@starr-apps:/opt/infisical# docker exec -it cross-seed sh
/config $ npm install dotenv
npm error code EACCES
npm error syscall mkdir
npm error path /.npm
npm error errno EACCES
npm error
npm error Your cache folder contains root-owned files, due to a bug in
npm error previous versions of npm which has since been addressed.
npm error
npm error To permanently fix this problem, please run:
npm error   sudo chown -R 3066:3066 "/.npm"
npm error Log files were not written due to an error writing to the directory: /.npm/_logs
npm error You can rerun the command with `--loglevel=verbose` to see the logs in your terminal
/config $
```

**Reply by mmgoodnow** [reply]
you don't need to use dotenv - should be able to just change the container entry point to `node --env-file /config/.env /path/to/dist/cmd.js`

**Reply by zakkarry** [reply]
Oh, I thought you meant you couldn't get a shell, not that permissions issues existed from inside the container.

I think it is unlikely that we'd add a package that isn't used by cross-seed to the container itself, but I can bring it up. There will perhaps be a bigger discussion around addressing getting env vars to work inside cross-seed again (we had it in initial v6 pre-release but removed it due to complications) over adding an auxiliary unused package, we'll be moving to 22 LTS soon and I think support for this is better there than 20.

You could always modify the entry point in compose or fork, set the environmental variables and use process.env in the config file, I'd think. Not entirely sure if this works exactly for your uses, but it would seem like an avenue to look into in the meantime.

Edit: Seems like @mmgoodnow had the same thought as I did.

**Reply by LilTrublMakr** [reply]
Makes sense about not adding it. I was thinking since it was a very small package, it could possibly be feasible.

Wouldn't messing with the entrypoint override the one in the dockerfile? Genuinely asking, I am not completely sure how docker works.

I have tried just passing through the environment file using `env_file: .env` in the compose file but my guess is that is broken from what zakkarry has said?

I have also tried to make a sh file and running it using `command: daemon && source /config/env.sh` and it does not seem to run when the container starts. I am able to get in to shell, run it manually, and it works. 

My overall goal is to use Infisical to automatically automate changing the API keys if they ever change. I recently suffered from a failure and it was a pain to track down where all the API keys were used to change them. It would definitely cut down the places I need to change them if I am able to script them in.

**Reply by zakkarry** [reply]
Our previous implementation used the same method that @mmgoodnow said, but the issue was when there was no env available it wouldn't run correctly, you wouldn't have this issue.

modifying the entrypoint in compose would indeed override the default, but that's what you want in this case and it would launch with the .env file loaded, and thus you would get the desired outcome.

**Reply by LilTrublMakr** [reply]
Alright. So `node --env-file /config/.env /path/to/dist/cmd.js` as the entrypoint but what would I replace `/path/to/dist/cmd.js` with? I am not seeing any .js in the current entrypoint of the dockerfile.

**Reply by zakkarry** [reply]
`/usr/local/lib/node_modules/cross-seed/dist` in our container has the .js files.

**Reply by LilTrublMakr** [reply]
Hmm... Might just have to curb this one until environment variables are fixed...

`entrypoint: node --env-file /config/.env /usr/local/lib/node_modules/cross-seed/dist/cmd.js` and then using 

```
var sonarr_apikey = process.env.SONARR_API;
var radarr_apikey = process.env.RADARR_API;
var prowlarr_apikey = process.env.PROWLARR_API;
```
in config.js and it seems like when going to do a manual search (docker exec cross-seed cross-seed search), the env vars were not passed through.

```
info: cross-seed v6.4.1
info: Validating your configuration...

error: Indexer http://prowlarr.domain.tld:9696/1/api responded with code 401 when fetching caps, check verbose logs
[truncated... ^ a few more times...]
info: [qbittorrent] Logged in to qBittorrent v5.0.3
error: Could not contact Sonarr at https://sonarr.domain.tld/?apikey=[REDACTED]
info: Validating all existing torrent save paths...
```

**Reply by zakkarry** [reply]
That's because the alias `cross-seed` does not pass the .env file, you'll need to run `cmd.js search` - its not loading them into environmental variables, it's loading the .env into node.

You can load them using the source command in a script if you want to, and use that as a preceding command to the normal entry commands (running daemon mode) - or you can maybe alter the alias...all `cross-seed <arg>` does is run cmd.js

**Reply by zakkarry** [reply]
There's no reason this can't work for you, you just have to make modifications to how you're running cross-seed to send the .env file, or actually load it as environmental variables. I'd opt for modifying the alias and passing the .env like you did in the entry point, personally.

You can set up the entry point with a script to do this in a way that becomes transparent after updates and recreation of the container, but you can't just run the entrypoint and expect everything else to be psychic, that's not how passing --env-file works

**Reply by LilTrublMakr** [reply]
Ah, that does make sense. So if I understand correctly, I should be able to make a script file that modifies the alias, use that script as the entrypoint, and then both the cron and manual searches should work? Would this also work for the searches called from API (torrent completed script)?

**Reply by zakkarry** [reply]
There is technically no alias for `cross-seed` - it is a symlink that is in `/usr/local/bin/` - so you'd need to create an alias I suppose. 

Running the command `alias cross-seed='/usr/local/bin/cross-seed --env-file /config/.env "$@"'` would create the alias for cross-seed and accept all the arguments, you can add in the --env-file argument before the $@ part. This assumes the path for the .env, but I think it demonstrates what you want if its somewhere else. I tested this alias command in my container and it functioned as intended.

This would need to run everytime the container started to ensure that it was present when the container was recreated (updates, etc) and then you can run cross-seed normally. So you'd want to make a .sh script that runs the alias command, and then proceeds to execute cross-seed with the argument provided to the entry point.

```sh
#!/bin/sh
alias cross-seed='node /usr/local/lib/node_modules/cross-seed/dist/cmd.js --env-file /config/.env "$@"'
cross-seed "$@"
```

The completed script sends a curl command to the API end point of the daemon, it doesn't require a separate process and as long gas you had the daemon running it would function with the .env variables present.

That should cover you, though. I haven't tested it in full, just the alias command and its functionality, but this should be enough to get you going with the entry point modification and the cross-seed command.

**Reply by LilTrublMakr** [reply]
Unfortunately still not working.

From my understanding, when you run a shell script, the alias will only work for that shell commands environment and duration. When the script exits, it does not carry over to the host machine. The way to get around this would be to use `source` or the dot command (`.`) to export it to the host machine. However, this does not seem to be sticking. I have tried MANY combinations of the entrypoint and just can not seem to get it to stick in the host machine without sh-ing in to the machine and running `. entrypoint.sh` manually. 

For anyone trying to do this in the future: I am able to use the entrypoint to run the daemon with the .env file using `entrypoint: node --env-file /config/.env /usr/local/lib/node_modules/cross-seed/dist/cmd.js daemon` in my compose file (assuming the `.env` file is in the config folder). 

To manually start a search, I am using `docker exec cross-seed node --env-file /config/.env /usr/local/lib/node_modules/cross-seed/dist/cmd.js search`. Again, assuming the `.env` file is in the config folder and the docker container is named `cross-seed`.

As much as I would love this to be automated and be able to use just `cross-seed`, I don't think I want to spend any more time on this. But a huge thank you to Zakkarry for helping me get this far and for the time they are putting in to this awesome project!

**Reply by zakkarry** [reply]
I think you are very close, and there's just a minor misunderstanding between us on how things are trying to get done.

If you are on IRC or Discord or anything, I think we could get you where you want to be in 10-15min at the most, but I completely understand if you want to put it down for a while.

From another angle, if you really want to (and because I appreciate your container and work to get it on unraid before we were, and then your willingness to work with us when we had our own) I could help you with a fork that I could show you how to keep up to date. I know that's not ideal either, but I very much would like to help you get to where you need to be.

If this sounds good to you now or in the future, then we can discuss either the way to get the current thing we were attempting or a fork for you with the packages you need. 

No matter which of those you choose, it's definitely not that big of a problem.

If you just need dotenv installed in the docker container, and that's all, let me know too. I can fork myself and push a docker image you can use (I already keep multiple branches up to date, so adding yours to the list for the time being would be easy...)

**Reply by mmgoodnow** [reply]
looks like Node added support for this in 22.9: https://nodejs.org/dist/v22.12.0/docs/api/cli.html#--env-file-if-existsconfig so we should be able to support this OOTB when we move to 22.

**Reply by zakkarry** [reply]
> looks like Node added support for this in 22.9: https://nodejs.org/dist/v22.12.0/docs/api/cli.html#--env-file-if-existsconfig so we should be able to support this OOTB when we move to 22.

That's great, that was the reason we had to remove it.

@LilTrublMakr https://github.com/cross-seed/cross-seed/compare/master...dotenv - you can switch to the `dotenv` tag - it's master/latest with dotenv added for you. 

I'm not sure if there was something else you needed or whatever, but if that was all I can keep this image relatively up to date (within a few hours of publishing usually, but sometimes it takes 12-24 hours for me to rebase everything, i can automate this fairly easily though for this branch) but if this is all, then hopefully this helps.

As @mmgoodnow said, once we are in node 22 minimum, we can add official support back in.

**Reply by zakkarry** [reply]
> Unfortunately still not working.
> 
> From my understanding, when you run a shell script, the alias will only work for that shell commands environment and duration. When the script exits, it does not carry over to the host machine. The way to get around this would be to use `source` or the dot command (`.`) to export it to the host machine. However, this does not seem to be sticking. I have tried MANY combinations of the entrypoint and just can not seem to get it to stick in the host machine without sh-ing in to the machine and running `. entrypoint.sh` manually.
> 
> For anyone trying to do this in the future: I am able to use the entrypoint to run the daemon with the .env file using `entrypoint: node --env-file /config/.env /usr/local/lib/node_modules/cross-seed/dist/cmd.js daemon` in my compose file (assuming the `.env` file is in the config folder).
> 
> To manually start a search, I am using `docker exec cross-seed node --env-file /config/.env /usr/local/lib/node_modules/cross-seed/dist/cmd.js search`. Again, assuming the `.env` file is in the config folder and the docker container is named `cross-seed`.
> 
> As much as I would love this to be automated and be able to use just `cross-seed`, I don't think I want to spend any more time on this. But a huge thank you to Zakkarry for helping me get this far and for the time they are putting in to this awesome project!

Did you try out that branch/tag with dotenv added?

**Reply by LilTrublMakr** [reply]
Sorry, I have not. I moved on to other projects (Currently trying to get Ansible to pull inventory from Netbox so I don't have to manually edit files when I add/remove a service). I can wait for 22 to come out to come back to it with native support.

**Reply by zakkarry** [reply]
Ok, I'll remove the branch then, and I'll let you know when we merge that #858 should be when we move to 22 I think
